designs/0030-complex-numbers.md

```diff
@@ -0,0 +1,328 @@
+
+The goal of this document is to be precise enough that
+    - users can understand the feature well enough to evaluate it, and
+    - developers can work on a technical specification for the implementation.
+
+## Design considerations
+
+- The main design consideration is that Stan programs involving complex numbers be easy to write, and just as importantly, easy to read.
+
+- We should follow the basic design of Stan types, including promotion structure and associated covariance (in the CS not stats sense).
+
+- I/O structures need to
+    - balance self documentation and efficiency, and
+    - be easy to adapt to the interfaces and analysis tools.
+
+
+## Assumptions
+
+- The language will target C++ code in the Stan math library.
+
+
+## Stan data types
+
+The following four basic data types will handle all of the complex number functionality we need:
+
+- `complex`:  scalar type
+- `complex_vector[N]`: vector of complex values
+- `complex_row_vector[M]`: row vector of complex values
+- `complex_matrix[M, N]`: matrix of complex values
+
+As usual, arrays can be constructed from any of these types.
+
+
+### Constructors
+
+A complex number is constructed from a pair of real numbers, one representing the real component and one the imaginary component.  The proposal is to have a function do that work.
+
+```
+real r;
+real i;
+complex z = to_complex(r, i);
+```
+
+This overloads the type name in the same way as a C++ constructor.  `to_complex` seems too wordy and `cmplx` too cryptic.
+
+### Imaginary literals
+
+Imaginary literals are formed the same way as in MATLAB, R, and Python using a real literal followed by a letter.  In Stan's case, we use `i`.  That means that `3i`, `-2.9i` or `157.23458i` are all complex numbers with the specified imaginary component and a zero real component.  These can be put together with addition to construct complex numbers using mathematical notation.
+
+```
+complex z = 3 - 2.9i;
+```
+
+### Accessors
+
+The real and imaginary components of a complex number may be accessed as follows.
+
+```
+complex z;
+...
+real r = get_real(z);
+real i = get_imag(z);
+```
+
+These accessors do not work as lvalues. 
+
+
+### Equality
+
+Equality for scalars is defined component-wise, so that if `z1` and `z2` are are complex typed variables, then `z1 == z2` is true if `to_rea(z1) == to_real(z2)` and `to_real(z1) == to_real(z2)`.
```

Comment:
```suggestion
Equality for scalars is defined component-wise, so that if `z1` and `z2` are are complex typed variables, then `z1 == z2` is true if `get_real(z1) == get_real(z2)` and `get_imag(z1) == get_imag(z2)`.
```